sklearn/manifold/isomap.py

```diff
@@ -45,19 +45,63 @@ class Isomap(BaseEstimator, TransformerMixin):
         Maximum number of iterations for the arpack solver.
         not used if eigen_solver == 'dense'.
 
-    path_method : string ['auto'|'FW'|'D']
-        Method to use in finding shortest path.
+    path_method : string ['auto'|'FW'|'D'], optional
+        Algorithm to use for shortest paths.  Options are:
+
+           'auto' -- (default) select the best among 'FW', 'D', 'BF', or 'J'
+                     based on the input data.
 
-        'auto' : attempt to choose the best algorithm automatically.
+           'FW'   -- Floyd-Warshall algorithm.  Computational cost is
+                     approximately ``O[N^3]``.  The input csgraph will be
+                     converted to a dense representation.
 
-        'FW' : Floyd-Warshall algorithm.
+           'D'    -- Dijkstra's algorithm with Fibonacci heaps.  Computational
+                     cost is approximately ``O[N(N*k + N*log(N))]``, where
+                     ``k`` is the average number of connected edges per node.
+                     The input csgraph will be converted to a csr
+                     representation.
 
-        'D' : Dijkstra's algorithm.
+           'BF'   -- Bellman-Ford algorithm.  This algorithm can be used when
+                     weights are negative.  If a negative cycle is encountered,
+                     an error will be raised.  Computational cost is
+                     approximately ``O[N(N^2 k)]``, where ``k`` is the average
+                     number of connected edges per node. The input csgraph will
+                     be converted to a csr representation.
+
+           'J'    -- Johnson's algorithm.  Like the Bellman-Ford algorithm,
+                     Johnson's algorithm is designed for use when the weights
+                     are negative.  It combines the Bellman-Ford algorithm
+                     with Dijkstra's algorithm for faster computation.
 
     neighbors_algorithm : string ['auto'|'brute'|'kd_tree'|'ball_tree']
         Algorithm to use for nearest neighbors search,
         passed to neighbors.NearestNeighbors instance.
 
+    n_landmarks : [None|int|'auto'] (default = None)
+        The number of landmarks. If this parameter was set,
+        L-Isomap (Landmark Isomap) will execute instead of the
+        original algorithm.
+
+        Should be a number sufficiently greater than n_components + 1, for
+        stability [2].
+
+        None : all samples will be used. The original Isomap algorithm will be
+        executed.
+
+        'auto' : automatically computes the number of landmarks to use and
+        randomly-selects them.
+
+        int :  use exactly n_landmarks randomly-selected landmarks.
+
+    landmarks : array-like, optional (default = None)
+        The list of landmarks to use. If none was passed,
+        infers it from n_landmarks.
```

Comment:
Rather than both ``n_landmarks`` and ``landmarks``, I think it would make sense to have a single parameter ``landmarks`` which is either ``None``, ``'auto'``, an integer (specifying number of landmarks), or an array-like (specifying the indices of the landmarks).